portable/MemMang/heap_4.c

```diff
@@ -183,8 +191,95 @@ void * pvPortMalloc( size_t xWantedSize )
                 pxPreviousBlock = &xStart;
                 pxBlock = xStart.pxNextFreeBlock;
 
-                while( ( pxBlock->xBlockSize < xWantedSize ) && ( pxBlock->pxNextFreeBlock != NULL ) )
+                while( ( pxBlock->pxNextFreeBlock != NULL ) )
                 {
+
+                    if ( pxBlock->xBlockSize < xWantedSize )
+                        goto nextFreeBlock;
+
+#ifdef __CHERI_PURE_CAPABILITY__
+
+                    /* The block will be already representable as portBYTE_ALIGNMENT is bigger than
+                     * the required alignment. Do nothing, found the block.
+                     */
+                    if ( xRepresentableAlignmentBytes <= portBYTE_ALIGNMENT )
+                        break;
+
+                    /* Get the start address of the free block */
+                    void* blockStartAddress = ( void * ) ( pxBlock );
+                    size_t xRepresentableAlignmentGap = 0;
+                    void* alignedStartBlockAddr = NULL;

```

Comment:
Arithmetic on `void *` is undefined. Use `char *` instead (will need to percolate through the rest of the file). Avoid gratuitous stores while here.
```suggestion
                    char* blockStartAddress = ( char * ) ( pxBlock );
                    size_t xRepresentableAlignmentGap;
                    char* alignedStartBlockAddr;
```